File: src/linkedlist/lc_206_reverse_linked_list/reverse_linked_list.py
from listNode import ListNode
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head

        p1 = head
        p2, p3 = None, None

        while p1:
            p2 = p1
            p1 = p1.next
            p2.next = p3
            p3 = p2
        head = p3
        return head
        

File: src/linkedlist/lc_206_reverse_linked_list/reverse_linked_list.java
class Solution {
    static class ListNode {
        int val;
         ListNode next;
         ListNode() {}
         ListNode(int val) { this.val = val; }
         ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    
    public ListNode reverseList(ListNode head) {
        ListNode temp1 = head;
        ListNode temp2;
        ListNode temp3 = null;

        while(temp1 != null) {
            temp2 = temp1;
            temp1 = temp1.next;
            temp2.next = temp3;
            temp3 = temp2;
        }
        return temp3;
    }
}